source/MaterialXTest/ShaderGen.cpp

```diff
@@ -40,6 +41,98 @@ bool isTopologicalOrder(const std::vector<const mx::SgNode*>& nodeOrder)
     return true;
 }
 
+// Observer to add library markers to all nodes. Use the "namespace" attribute to do this
+static std::string NAMESPACE_ATTRIBUTE("namespace");
+class LibraryObserver : public MaterialX::Observer
+{
+public:
+    LibraryObserver() {}
+
+    void onAddElement(MaterialX::ElementPtr /*parent*/, MaterialX::ElementPtr element) override
+    {
+        if (_libraryRefName.length())
+        {
+            element->setAttribute(NAMESPACE_ATTRIBUTE, _libraryRefName);
+        }
+    }
+
+    void setLibraryRefName(const std::string& name)
+    {
+        _libraryRefName = name;
+    }
+
+protected:
+    // Libraryref name to add
+    std::string _libraryRefName;
+};
+
+//
+// Get source content, source path and resolved paths for
+// an implementation
+//
+bool getShaderSource(MaterialX::ShaderGeneratorPtr generator,
+                    const MaterialX::ImplementationPtr implementation,
+                    MaterialX::FilePath& sourcePath,
+                    MaterialX::FilePath& resolvedPath,
+                    std::string& sourceContents) 
+{
+    if (implementation)
+    {
+        sourcePath = implementation->getFile();
+        resolvedPath = generator->findSourceCode(sourcePath);
+        if (MaterialX::readFile(resolvedPath.asString(), sourceContents))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+//
+// Find the implementation for specific language or target for a given Element
+//
+// Should this be moved under the document class?
+MaterialX::ImplementationPtr findImplementation(mx::DocumentPtr document, const std::string& name,
+                                                const std::string& language, const std::string& target)
+{
+    std::vector<MaterialX::InterfaceElementPtr> impllist = document->getMatchingImplementations(name);
+    for (auto elem : impllist)
+    {
+        MaterialX::ImplementationPtr implementation = elem->asA<MaterialX::Implementation>();
```

Comment:
The implementation can also be a NodeGraph element, so need to check for that here. You can use `implementation->isA<NodeGraph>()`

Check how it's node here: https://github.com/autodesk-forks/MaterialX/blob/adsk_update_shadergen/source/MaterialXShaderGen/SgNode.cpp#L134

Reply:
I wasn't thinking at this level, as nodegraph implementations can't really "fail to exist" but will add since it makes the utility complete. With this, the same question here is if this one would belong somewhere in core? Can decide later naturally.